api/v1alpha1/addresspool_types.go

```diff
@@ -0,0 +1,81 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type AddressPoolType struct {
+	// Address Pool Name
+	Name string `json:"name"`
+
+	// Protocol can be used to select how the announcement is done,
+	// +kubebuilder:validation:Enum:=layer2; bgp
+	Protocol string `json:"protocol"`
```

Comment:
I find it a little odd to have `protocol` as part of the `AddressPool`.  I know that's how the data is structured in the ConfigMap, but as best I can tell, MetalLB only supports speaking one protocol at a time.  You can't actually have address pools of different types at the same time.

How about splitting this out into a CRD for global MetalLB configuration options?  `MetalLBConfig` or something ...

Reply:
we discussed that in upstream design discussion and I agree to it multiple suggestions here 
- fix it in this PR ( I don't prefer given the size of this PR)
- open issue for it and address in the upcoming PR
- defer to feature release in case if we find use case of with future protocols combo that might be ok to mix for example if we ever supported OSPF and BGP in that case its prefectly fine to have ?

Reply:
I don't mind if it's addressed in a future PR, but we should not defer to a future release, because it would be a breaking API change.

I'd also be surprised if we ever mixed routing protocols in the future.  That's just my guess though ...

Reply:
> I don't mind if it's addressed in a future PR, but we should not defer to a future release, because it would be a breaking API change.
> 
> I'd also be surprised if we ever mixed routing protocols in the future. That's just my guess though ...

as discussed offline we will check if same protocol used on multiple objects will fail the config via webhook otherwise we will let user configure different protocol as its possible to find future use cases when FRR is in the mix for example mixing BGP and OSPF , I created #21 to follow up on this Thanks!!